### ISSUE TITLE
add qx translation

### ISSUE THREAD

**Issue by johnspackman** [qooxdoo/qooxdoo-compiler#85]
_From @hkollmann on November 30, 2017 20:3_

to extract *.po files.
Similar to generate translation

_Copied from original issue: qooxdoo/qooxdoo-cli#56_

**Comment by johnspackman** [comment]
_From @cajus on November 30, 2017 20:4_

That is already done automatically. The .po files are extracted and updated by the compiler.

**Comment by johnspackman** [comment]
_From @hkollmann on November 30, 2017 20:57_

Then there is missing some doc. 

qx create test -I
cd test

add "de" to locales:  "locales": ["en", "de"],

run qx compile

i can't find any po file?

**Comment by johnspackman** [comment]
_From @cboulanger on November 30, 2017 21:0_

see https://github.com/qooxdoo/qooxdoo-cli/issues/38#issuecomment-346944443

**Comment by johnspackman** [comment]
You need to be calling `this.tr("myMessageId")`, that's how the compiler finds the message and know to generate a translation for it

**Comment by johnspackman** [comment]
_From @hkollmann on November 30, 2017 21:18_

ok, got it. I'll change the text in readme.txt of the skeletons

**Comment by johnspackman** [comment]
_From @cboulanger on December 1, 2017 7:13_

@johnspackman has noted that "de", "en" etc. won't work anyways because the compiler requires "real" locale codes such as "de_DE" etc. 

a) This seems to be not BC with existing applications, which use only two-letter locales
b) we need to validate locales in `qx create` and ideally, use the same locale data that qxcompiler uses to validate entries. @johnspackman, is there a way to use the qxcompiler API to get this data easily?

**Comment by johnspackman** [comment]
_From @cajus on December 1, 2017 7:16_

@cboulanger existing applications use *both*, two-letter and combined locale definitions. de is for de* (i.e. it catches de-AT and de-DE if they don't exist.

**Comment by johnspackman** [comment]
_From @cboulanger on December 1, 2017 7:18_

@cajus Ah ok, thank you. I didn't know that.

**Comment by johnspackman** [comment]
_From @cajus on December 1, 2017 7:19_

Just checked with an existing one. Current correct spelling would be *en_US* for american english. So dash vs underline.

**Comment by johnspackman** [comment]
_From @cboulanger on December 1, 2017 7:20_

But won't two-letter code apps break with the new compiler, @johnspackman? If so we need to remember to provide an upgrade.

**Comment by johnspackman** [comment]
_From @cajus on December 1, 2017 7:22_

I don't know the state of the *Part* support in qx-compiler, but that's also a must have:

`"i18n-as-parts"    : true`

is present in the current generator. It can bundle locales as parts - which makes sense if you've 25 locales with ~200k each.

**Comment by johnspackman** [comment]
@cboulanger No, what happens is that if you just specify the language you will only get the default localisation settings for that language; this means that "en" will always return a date format "m/d/y" because the defaults happen to be US-centric (for dates anyway).  It works but it's wrong (unless you happen to be in the US).  If you specify "en_GB", the GB localisation data for "en" is loaded on top of the "en" data, overriding those "en" defaults.

It means though that you have to know in advance which language/locale combinations you want to support, so that the generator/compiler knows to extract the data it requires and insert it into the boot loader.

@cajus the `i18n-as-parts` is not in compiler at the moment, however it only loads a tiny subset of locale data is provided with the app - the entire "en" language is 4,698 bytes (pretty printed, un-minimized JSON data), and the "en_GB" specialization is another 1,858 bytes.  When minimized, "en" and "en_GB" combined come to 4,782 bytes.

**Comment by johnspackman** [comment]
_From @cajus on December 1, 2017 7:41_

@johnspackman yes, but if you *have* large translations, it's a problem. Or at least a waste of resources. For one application I've more than 4MB of translations. If this is pulled in at a time, 1/25 is used. 24/25 are wasted and would be loaded.

@hkollmann just deleted my *de.po* and it's getting re-generated upon `qx compile`. The contents is not valid in the moment, because *qooxdoo-compiler* needs an update on npm AFAIK.

**Comment by johnspackman** [comment]
ah, yes I see what you mean!  That's not supported yet but it needs to be. I'll create an issue

**Comment by johnspackman** [comment]
_From @cajus on December 1, 2017 7:59_

Maybe it would be nice to have a seperate command to extract the translations (i.e. `qx translate`). Most of the time, you don't want `qx compile` make that.

**Comment by cboulanger** [comment]
This should be decided before going into beta. Should `qx compile` update translations or a special command? See also #161 and #163 

CC: @hkollmann @oetiker

**Comment by oetiker** [comment]
ideally it would ONLY update if there are altered strings ... or if explicitly called for ... at the moment it updates all the time and the changes are mostly line numbers in the comments.

**Comment by cboulanger** [comment]
+1 for updating only if translations have changed.

**Comment by johnspackman** [comment]
it's not really possible to not update the line numbers without breaking the purpose of having line numbers in there, and the file format is the [GNU gettext](https://www.gnu.org/software/gettext/manual/html_node/PO-Files.html) format.

Our options are to avoid line numbers altogether or to maybe use sourcemaps - but then that depends on an editor which supports source maps when editing gettext .po files.

This is made much worse by a bug which outputs loads of the wrong references - that's now fixed (see next PR)

So I suggest that with these bug fixes in conjunction with only updating .po on demand, we leave it for now and see how it works in practice.  If it's still a PITA for git commits, maybe we could remove the line numbers?

**Comment by oetiker** [comment]
@johnspackman my idea is sort of a middle ground ... only update the files if the actual translations change ... yes, the line numbers may be a bit out of sync on times, but it would cut down on the changes ... and even slightly out of date line numbers might be better than no line numbers at all ...

**Comment by johnspackman** [comment]
How about we just remove the line number altogether?  The `msgid` is likely to be unique and fairly easy to search for inside a source file, and so long as the list of filenames is accurate and consistent that should be enough to locate all instances (without providing wrong information) and not pad out commits?

**Comment by cboulanger** [comment]
Maybe the current solution is good enough for the moment? It is not too much to ask the developer to manually update the translations after working on them? A more convenient solution could be developed later.

**Comment by cboulanger** [comment]
I never had use for  the line numbers. As far as I am concerned, they could be omitted.

**Comment by oetiker** [comment]
I never use the line numbers either ... fine with me to drop them ... only to update translations on special requests is not really ideal in my eyes :)

**Comment by zaucker** [comment]
This actually would be my preferred solution from a git workflow perspective
as it would only require committing changes that are actually of potential
interest.

I am not sure if this would follow .po file specifications (and thus could
break some tools relying on the line number). Which for me would not be a 
problem ;-)

On Wed, 1 May 2019, John Spackman wrote:

> How about we just remove the line number altogether?  The `msgid` is
> likely to be unique and fairly easy to search for inside a source file,
> and so long as the list of filenames is accurate and consistent that
> should be enough to locate all instances (without providing wrong
> information) and not pad out commits?

-- 
Oetiker+Partner AG              tel: +41 62 775 9903 (direct)
Fritz Zaucker                        +41 62 775 9900 (switch board)
Aarweg 15                            +41 79 675 0630 (mobile)
CH-4600 Olten                   fax: +41 62 775 9905
Schweiz                         web: www.oetiker.ch

**Comment by cboulanger** [comment]
What about adding line numbers on demand only or by a qx config setting?

**Comment by johnspackman** [comment]
yes, a `qx.strictPoCompatibility` setting which is false by default

**Comment by cboulanger** [comment]
+1 but maybe call it qx.translation.strictPoCompatibility to have a namespace for other translation settings
